### ISSUE TITLE
Wrong type of "this" pointer detected in a non-const method

### ISSUE THREAD

**Issue by Leedehai** [microsoft/vscode-cpptools#2303]
Type: LanguageService

**Describe the bug**
- OS and Version: macOS High Sierra 10.13.5
- VS Code Version: 1.25.1
- C/C++ Extension Version: 0.17.7
- Other extensions you installed (and if the issue persists after disabling them): only this extension pertains to C/C++.
- A clear and concise description of what the bug is: In a non-const, no-volatile method of class `MyClass`, the `this` pointer should be of type `MyClass *` in the hovering message (per C++11 standard 9.3.2 paragraph 1, C++17 standard 12.2.2.1 paragraph 1), **not** `MyClass *const`.

**To Reproduce**
1. Go to VSCode, open a fresh new file `test.cc`.
2. Type in the code below.
```
class MyClass {
public:
	MyClass *getPointer() { return this; }
};
int main() {}
```
3. Save it.
4. Hover above the `this` pointer, examine the message.
5. The message says it's `MyClass *const this`.

**Expected behavior**
The type message should be `MyClass *this`.

**Screenshots**
![1081532133048_ pic](https://user-images.githubusercontent.com/18319900/43030269-fd8eace0-8c42-11e8-98fb-7c21f466fcee.jpg)

**Additional text**: this [StackOverflow](https://stackoverflow.com/a/6067267/8385554) post said in the days of C++98 and C++03 old GCC and MSVC compilers internally used a trick that labels the `this` pointer as a const pointer. That trick was benign at that time but became problematic in C++11 and later standards. GCC has since abandoned that trick. I'm wondering if this is the problem. That post also said as of 2017 MSVC didn't fix the problem.

**Comment by sean-mcmanus** [comment]
I believe this is "by design" and repros in VS 2017. The type "MyClass *const" is correct for "this" and it means "this" cannot be assigned to (it's const) and it points to an object of type "MyClass *", which is not const. If getPointer() was defined as "MyClass *getPointer() const" then the type of "this" would be "const MyClass *const".

**Comment by Leedehai** [comment]
@sean-mcmanus But it violates the C++11 standard :( so there is currently no plan in the VS team to be more aligned with the standard on this issue? And, does VSCode's cpptool use VS as its internal engine?

**Comment by sean-mcmanus** [comment]
Okay, it looks like you're right, it's a bug: https://stackoverflow.com/questions/6067244/type-of-this-pointer .

It got reported to VS: https://developercommunity.visualstudio.com/content/problem/293655/class-foo-const-this-reported-while-in-not-const-c.html . If they fix it, we'll take the fix as well.

**Comment by sean-mcmanus** [comment]
VS "Won't Fixed" the bug saying it didn't have enough "votes", but when I tried to upvote it, it wouldn't let me, so I created a new issue at https://developercommunity.visualstudio.com/content/problem/416906/c-intellisense-wrong-type-of-this-pointer-detected.html -- that issue will probably need to get enough upvotes in order to get fixed though.

**Comment by sean-mcmanus** [comment]
Fixed with 0.25.0.
